debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbSnapshotChangeEventSource.java

```diff
@@ -0,0 +1,599 @@
+                return SnapshotResult.aborted();
+            }
+
+            snapshotProgressListener.snapshotCompleted();
+
+            return SnapshotResult.completed(ctx.offset);
+        }
+        catch (InterruptedException e) {
+            LOGGER.warn("Snapshot was interrupted before completion");
+            snapshotProgressListener.snapshotAborted();
+            throw e;
+        }
+        catch (RuntimeException e) {
+            snapshotProgressListener.snapshotAborted();
+            throw e;
+        }
+        catch (Throwable t) {
+            snapshotProgressListener.snapshotAborted();
+            throw new RuntimeException(t);
+        }
+        finally {
+            LOGGER.info("Snapshot step 4 - Finalizing");
+            complete(ctx);
+        }
+    }
+
+    protected SnapshottingTask getSnapshottingTask(OffsetContext previousOffset, ReplicaSets replicaSets) {
+        if (previousOffset == null) {
+            LOGGER.info("No previous offset has been found");
+            if (connectorConfig.getSnapshotMode().equals(MongoDbConnectorConfig.SnapshotMode.NEVER)) {
+                LOGGER.info("According to the connector configuration, no snapshot will occur.");
+                return new SnapshottingTask(Collections.emptyList());
+            }
+            return new SnapshottingTask(replicaSets.all());
+        }
+
+        // Even if there are previous offsets, if no snapshot should occur, return task with no replica sets
+        if (connectorConfig.getSnapshotMode().equals(MongoDbConnectorConfig.SnapshotMode.NEVER)) {
+            LOGGER.info("According to the connector configuration, no snapshot will occur.");
+            return new SnapshottingTask(Collections.emptyList());
+        }
+
+        // Collect which replica-sets require being snapshotted
+        final List<ReplicaSet> replicaSetSnapshots = new ArrayList<>();
+        final MongoDbOffsetContext offsetContext = (MongoDbOffsetContext) previousOffset;
+        try {
+            replicaSets.onEachReplicaSet(replicaSet -> {
+                MongoPrimary primary = null;
+                try {
+                    primary = establishConnectionToPrimary(replicaSet);
+                    final ReplicaSetOffsetContext rsOffsetContext = offsetContext.getReplicaSetOffsetContext(replicaSet);
+                    if (primary != null && isInitialSyncExpected(primary, rsOffsetContext)) {
+                        replicaSetSnapshots.add(replicaSet);
+                    }
+                }
+                finally {
+                    if (primary != null) {
+                        primary.stop();
+                    }
+                }
+            });
+        }
+        finally {
+            taskContext.getConnectionContext().shutdown();
+        }
+
+        return new SnapshottingTask(replicaSetSnapshots);
+    }
+
+    private void delaySnapshotIfNeeded(ChangeEventSourceContext context) throws InterruptedException {
```

Comment:
There should be a common parent for `RelationalSnapshotChangeEventSource` and `MongoDbSnapshotChangeEventSource` that would contain a shared code like this.

Reply:
See commit 0158b56.